rfcs/text/refactoring-date-component.md

```diff
@@ -0,0 +1,407 @@
+
+    const handleChange = (event) => {
+      const { formattedValue, rawValue } = event.target.value;
+      setValue(formattedValue);
+      setHiddenValue(formatToIso(rawValue));
+    };
+
+    const handleBlur = (event) => { 
+      // validation can be run here
+     }
+
+    return (
+      <>
+        <Date
+          value={value}
+          onChange={handleChange}
+          onBlur={handleBlur}
+          name="date-input"
+          picker={(pickerProps) => <Picker {...pickerProps}/>}
+        />
+        <input
+          type="hidden"
+          name="date-input"
+          data-element="hidden-input"
+          value={hiddenValue}
+        />
+      </>
+    );
+  };
+```
+
+## DatePicker component
+
+We will continue to use the same sub-components such as `Navbar` and `Weekday`, although they should be moved to an `__internal__` directory of the `DatePicker` component as they will never need to be directly imported by consuming projects. The `containerProps` prop provides an interface for implementing keyboard accessibility within the picker (`tabIndex`, `onKeyDown` and so on). The [locale interface](https://react-day-picker.js.org/docs/localization) surfaced by `react-day-picker` will be integrated with `date-fns/locale` to override the translations strings. Below is an example of how translating weekdays can be achieved using the locales surfaced from `date-fns/locale`. It is also possible to generate an array of `months` based on the locale string.
+
+```js
+  const renderWeekdayElement = (weekdayElementProps) => {
+    const { className, weekday } = weekdayElementProps;
+    const weekdayLong = localize.day(weekday);     
+    const weekdayShort = localize.day(weekday, { width: "abbreviated" });
+    
+    return (
+      <Weekday className={className} title={weekdayLong}>
+        <abbr title={weekdayLong}>{weekdayShort}</abbr>
+      </Weekday>
+    );
+  }}
+```
+
+```js
+  const months = Array.from({ length: 12 }).map((_, i) => localize.month(i))
+```
+
+As with the existing component we will store a `lastValidValue` in internal state and use that to control the date displayed in the `DatePicker`. This can be achieved by leveraging a `useEffect` hook to update the state value if a new `inputValue` parses to a valid Date. We will also keep the same implementation the current component has with regards to rendering the `DatePicker` in a `portal` using the existing `Popover` component. To support additional features we will expose the `react-day-picker` interface via the `pickerProps`. For example, the following [feature request](https://github.com/Sage/carbon/issues/4275) can be supported this by utilising the [captionElement](https://react-day-picker.js.org/examples/elements-year-navigation) render prop, although we would implement this via a boolean prop and handle the the composition internally ustilising Carbon's existing `Select` component: below is an example of how this can be achieved.
```

Comment:
If the user enters an invalid date I think we should reset the `react-day-picker` state, it shouldn't show a selection.

Reply:
Yeah that's fair enough, I didn't really have a strong opinion on how this should behave so went with how our current component behaves. It's easy enough to add so I'm happy to implement it as it could help avoid some weird behaviour